### PR TITLE
feat: Scoped Storage Migration Service, Notification & Deck Picker Implementation

### PR DESCRIPTION
**Pull request by david-allison**
## Purpose / Description
This Pull Request implements the foreground service which performs the Scoped Storage Migration, and hooks it up to the Deck Picker

## Review Goals
* Ensure alpha users will not be placed in a bad position by this. This PR moves us to 'developers can test the migration', I'd rather keep this as too unstable for alpha, despite the amount of testing, given the impact.
* Question the `onStartCommand` isStarted logic
* Determine the TODOs which need to be moved to the Plan

This also contains a couple refactorings which could be split out if requested in the review.

## Fixes
Related #5304

## Approach
* Define a Service
* Define code to run a repeated task
* Define a Service Binder to abstract out the intricacies from DeckPicker
* Define an API which determines the state of the migration

## How Has This Been Tested?
⚠️ This needs more thorough testing. I'm submitting this PR to allow for this to occur. Once this testing is complete, we can allow 'GitHub alpha' users to use it.
* Tested on my test collection. All media transferred.

## Known Issues
* `Reviewer` is not tied in to the migration process yet
* `Media` is not blocked

## Checklist
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/services/MigrationService.kt [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+            }
+            notification?.notifyCompletion(result)
+
+            // display a toast to the user.
+            displayMigrationCompleted(result)
+            stopSelf()
+        }
+
+        private fun displayMigrationCompleted(result: Boolean) {
+            // TODO: This should be discussed
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/services/MigrationService.kt:110]: I don't like toasts, specially when outside the app. A heads-up notification or a default one would be better IMO

**Review by BrayanDSO — CHANGES_REQUESTED** [review]
About the overall process:
* My AnkiDroid directory still exists after the migration with the "essential files". Shouldn't they be deleted as well?
* I can't see the progress bar here, and the progress report string shows `0.00MB` for the total
    * Tested on my phone, Samsung Galaxy Note 10 Lite, Android 13, One UI 5
    * https://user-images.githubusercontent.com/69634269/214580402-3f2c1aed-f680-436f-aee6-d6abb752c27a.mp4

**Comment by BrayanDSO** [comment]
Oh, and why have you removed the `Strings` label?

**Comment by criticalAY** [comment]
@BrayanDSO I am the one working with the progress bar that is to be shown during migration and the UI element, Basically STOR-04 and STOR-05 so will pull the updates and then push the PR after this PR gets merged

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt [hunk 2]
```diff
@@ -2794,16 +2874,26 @@ open class DeckPicker :
...
+        Timber.i("Starting Migrate User Data Service")
+        migrationService.startForeground(this, MigrationService::class.java)
     }
 
     /**
      * Show a dialog that explains no sync can occur during migration.
      */
     private fun warnNoSyncDuringMigration() {
-        // TODO: Fetch and display real numbers
+        // TODO: handle value updates
```
[the remark this answers is not in the document]
Comment (criticalAY) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:2886]: yes I was waiting for it, I have that part ready at my end

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt [hunk 3]
```diff
@@ -2779,13 +2852,20 @@ open class DeckPicker :
...
+                        updateDeckList()
+                        handleStartup()
+                        startMigrateUserDataService()
                     }
                 }
             }
             if (elapsedMillisDuringEssentialFilesMigration > 800) {
                 showSnackbar(R.string.migration_part_1_done_resume)
             }
+            showSnackbar(R.string.migration_part_1_done_resume)
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:2868]: why this?

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt [hunk 4]
```diff
@@ -2779,13 +2852,20 @@ open class DeckPicker :
...
                 withProgress(getString(R.string.start_migration_progress_message)) {
                     withContext(Dispatchers.IO) {
-                        migrateEssentialFiles(baseContext)
+                        loadDeckCounts?.cancel()
+                        CollectionHelper.instance.closeCollection(false, "migration to scoped storage")
+                        ScopedStorageService.migrateEssentialFiles(baseContext)
+
+                        updateDeckList()
+                        handleStartup()
+                        startMigrateUserDataService()
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:2861]: Isn't this called below as well?

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt [hunk 5]
```diff
@@ -2794,16 +2874,26 @@ open class DeckPicker :
...
+        val text = migrationService.instance?.let { service ->
+            return@let service.totalToTransfer?.let { totalToTransfer ->
+                "\n\n" + getString(R.string.migration_transferred_size, service.currentProgress.toMB().toFloat(), totalToTransfer.toMB().toFloat())
+            }
+        }
+        // TODO: maybe handle onStorageMigrationCompleted()
+        // TODO: sync_impossible_during_migration needs changing
         MaterialDialog(this).show {
-            message(text = resources.getString(R.string.sync_impossible_during_migration, 5))
+            message(text = resources.getString(R.string.sync_impossible_during_migration, 5) + text)
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:2896]: Concatenating here seems wrong, but it should be temporary anyway

**Comment by BrayanDSO** [comment]
Forgot to reply about your points:
> Ensure alpha users will not be placed in a bad position by this. This PR moves us to 'developers can test the migration', I'd rather keep this as too unstable for alpha, despite the amount of testing, given the impact.

I think it is stable enough to alpha testers, and they can always refure to migrate or keep using an older version

> Question the onStartCommand isStarted logic

Seems correct to me.

> Determine the TODOs which need to be moved to the Plan

- Dialog offering to migrate on startup
    - After first install
    - After postpone
 - Better migration complete message
     - My preference is snackbar if on-app + notification out of the app, but toast instead of snackbar is acceptable inside the app

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt (hunk 5 above)
Comment (criticalAY) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:2896]: This one is temporary indeed

File: AnkiDroid/src/main/java/com/ichi2/anki/services/MigrationService.kt (hunk 1 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/services/MigrationService.kt:110]: Do we have a UI element like this currently? Could we extract this to a new issue?

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt (hunk 3 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:2868]: Bad conflict resolution, great catch, thanks!

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt (hunk 5 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:2896]: Agreed. Temp and I'm fine to leave this in. I won't 'resolve'.

File: AnkiDroid/src/main/java/com/ichi2/anki/services/MigrationService.kt (hunk 1 above)
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/services/MigrationService.kt:109]: > Do we have a UI element like this currently? 

I don't think so.

> Could we extract this to a new issue?

It should. Probably better than discussing a TODO here

Comment (criticalAY) [reply AnkiDroid/src/main/java/com/ichi2/anki/services/MigrationService.kt:109]: #13131 is there to handle it

**Comment by david-allison** [comment]
Fixed the issue with remaining being 0: `getFileSize` was the culprit (`conflict` didn't exist, AND it threw if we tried to get the directory size of an empty directory)
 
Screenshot 2023-01-31 at 00 12 12

**Comment by david-allison** [comment]
> My AnkiDroid directory still exists after the migration with the "essential files". Shouldn't they be deleted as well?

* We want to keep the collection as a backup
* We want to add a text file explaining the migration for advanced users

File: AnkiDroid/src/main/res/values/01-core.xml [hunk 6]
```diff
@@ -235,4 +235,10 @@
         comment="Generic message about an unknown error, to be shown in snackbars and such"
         translatable="false"
         >@string/webview_crash_unknown</string>
+
+    <!-- Scoped Storage Migration -->
+    <string name="migrating_data_message">Migrating your data</string>
+    <string name="migration_successful_message">Migration successful</string>
+    <string name="migration_failed_message">Migration failed</string>
+    <string name="migration_transferred_size">Migrated %1$.2f MB of %2$.2f MB</string>
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/res/values/01-core.xml:243]: I'd either remove this string (my preference) as I find the progress bar enough, which would be one less string to translate and we don't have much time to get it translated on most languages

OR remove the decimals after the numbers, because they are always `.00`

Implementer's choice

**Review by BrayanDSO — APPROVED** [review]
LGTM.

Left an implementer's choice comment

**Review by criticalAY — APPROVED** [review]

File: AnkiDroid/src/main/res/values/01-core.xml (hunk 6 above)
Comment (david-allison) [reply AnkiDroid/src/main/res/values/01-core.xml:243]: I feel for larger medical decks it is worthwhile. I'll remove the `.2`

Comment (david-allison) [reply AnkiDroid/src/main/res/values/01-core.xml]: done

**Comment by mikehardy** [comment]
Can someone please explain the expected testing flow? I did this:

prep1) `git co v2.15.6`
prep2) (make sure I am on JDK that is compatible with the old version, for me that is running `jdk11`, normally I'm testing higher versions...)
prep3) `./gradlew clean installPlayDebug --rerun-tasks` (the clean and --rerun-tasks are superstitious when I'm testing, if caching is working correctly in gradle they are unnecessary extra compile/build work)
prep4) sync my ankiweb test account so I have a functioning pre-historic AnkiDroid running with data in /sdcard/AnkiDroid

Now I try to test:

test0) make sure you have the github command line utility installed, because it's useful, it is called `gh`
test1) `gh pr checkout 13152`
test2) `./gradlew clean installPlayDebug --rerun-tasks` (again, I'm superstitious about clean builds)
test3) I have two AnkiDroid apps installed because one has package name suffix `.debug` now, okay, click that one
test4) ...nothing? I allowed it access to storage and I have an empty deck list. It doesn't see my old collection

If I sync etc, the collection is already scoped storage so I'm not offered any migration.

:thinking:

**Comment by david-allison** [comment]
`com.ichi2.anki.debug` has been introduced to ensure that development-based activities will not uninstall `com.ichi2.anki` (equivalent to a data wipe).

- https://github.com/ankidroid/Anki-Android/pull/13089

----

A new collection is now initialised at `/storage/emulated/0/Android/data/com.ichi2.anki.debug/files`. This does not require a storage update.

- https://github.com/ankidroid/Anki-Android/pull/13136
  - pending: https://github.com/ankidroid/Anki-Android/issues/13169

An existing user would have a collection at `/storage/emulated/0/AnkiDroid`. This requires a storage upgrade, and the UI to do so appears

**note** Since changes to `main` affect alpha testers, and we want this functionality to be opt-in while additional testing rakes place, this UI is currently optional (a button to migrate is provided).

We have not yet moved to API 30, so we can manually set the collection location to a legacy location (`/storage/emulated/0/AnkiDroid`). A collection will be created, and the migration can be tested.

**Comment by mikehardy** [comment]
Hmm - what I thought was strange was that I *had* a collection there in /sdcard/AnkiDroid - why didn't the app (.debug suffix or not...) see it and use it? Even if the migration was optional this seems like the prime testing flow doesn't it? (start with 2.15.6 synced with ankiweb, install this, check it? It was the most "happy path" thing I could think of...)

I'll try with a manual storage location setting as new step test4 and current step test4 will become test5

Thanks!

**Comment by mikehardy** [comment]
Okay that worked, but here's another testing note, that is not fixable really: Don't do this on a Play Services emulator or you can't use `su` to go into inaccessible areas of the filesystem (that is, the migrated scoped storage location...) which is obvious unless you forget :laughing: 

Anyway, it seemed to work fine. A couple UX notes:

1- The migrate button on deck picker stayed there post-migration until deck picker was made to UI refresh (app background + foreground did it for me)
2- When the migration is really really fast (I did it on a small collection first) I get the toast that says I can keep using AnkiDroid but there is no notification about the migration (too fast!) and no indication that it finished (see item 1) so I was a little confused if it was still running or not

But it worked it seems! More testing in a bit on a non-play services emulator and with a large collection, and I'll read through the code etc as well.

**Comment by BrayanDSO** [comment]
> why didn't the app (.debug suffix or not...) see it and use it?

New installs default to the in-app directory (e.g. storage/0/emulated/com.ichi2.anki.debug/files). If you want to make new installs default to the legacy storage, add `legacy_storage=true` to your `local.properties` file

**Comment by BrayanDSO** [comment]
> When the migration is really really fast (I did it on a small collection first) I get the toast that says I can keep using AnkiDroid

The snackbar shouldn't be shown if the migration is too fast (800ms). Apparently this PR breaks that. 

> and no indication that it finished (see item 1) so I was a little confused if it was still running or not

#13131

**Comment by criticalAY** [comment]
> > When the migration is really really fast (I did it on a small collection first) I get the toast that says I can keep using AnkiDroid
> 
> The snackbar shouldn't be shown if the essential files migration is too fast (800ms). Apparently this PR breaks that. Good catch
> 
> > and no indication that it finished (see item 1) so I was a little confused if it was still running or not
> 
> #13131

Yes waiting for this to be merged then will implement and push

**Comment by mikehardy** [comment]
Great - thanks and sorry if I'm duplicating already known stuff. I'm attempting to help with review + testing on this one but I'm catching up on progress and have obviously missed huge amounts of work + tracking, so I'm sadly duplicating some comments. Hopefully it's still a net positive :-)

**Review by mikehardy — APPROVED** [review]
This all looks good on a first and second scan, and it didn't crash or anything when I ran it locally - I'd just be slowing it down further by making it wait more. Let's go! Also, of course: this is awesome as a chunk of progress, really excited

**Comment by mikehardy** [comment]
@criticalAY it seemed like you had some work queued up behind this? I'm going to do a string sync now but hopefully you're unblocked 🚂

**Comment by criticalAY** [comment]
> @criticalAY it seemed like you had some work queued up behind this? I'm going to do a string sync now but hopefully you're unblocked 🚂 

Yep! Thankyou

File: AnkiDroid/src/main/java/com/ichi2/anki/services/ServiceConnection.kt [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+import timber.log.Timber
+
+/**
+ * Simplifies the interface for binding to a running service
+ * Access the service via [instance].
+ *
+ *
+ * @param T The service to encapsulate. Note: service's [Service.onBind] must return a [SimpleBinder]
+ */
+abstract class ServiceConnection<T : Service> {
```
Comment (oakkitten) [inline AnkiDroid/src/main/java/com/ichi2/anki/services/ServiceConnection.kt:33]: Nit: I'm not entirely sure about this.

When I had to use a 1 activity + 1 service combination, I finally settled on a simplified approach of putting the service into a static variable (Java times). No binders, no boilerplate. Coupled with an event bus (actually `EventBus`) approach (now obsolete but yeah), it was, I think, an approach that should be still valid today with some reasonable modifications.

To be fair, on of my gripes was the async nature of `onServiceConnected` IIRC. I wanted that to be sync. So I chose other options that allowed me to do more seamless animations and stuff. This might be irrelevant for AnkiDroid and my memories might be obsolete, so this is no more than food for thought

The bottom line here is, I think this entire class might be an overcomplication. I see that it's generic and is supposed to simplify things, but can't we get away with just not using a connection, or using one that merely forwards to the service object itself?

File: AnkiDroid/src/main/java/com/ichi2/anki/services/ServiceConnection.kt [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        override fun onServiceDisconnected(name: ComponentName) {
+            Timber.d("%s disconnected", name.shortClassName)
+            instance?.let {
+                onServiceDisconnected(it)
+                instance = null
+            }
+        }
+    }
+
+    fun startForeground(context: Context, clazz: Class<T>) {
```
Comment (oakkitten) [inline AnkiDroid/src/main/java/com/ichi2/anki/services/ServiceConnection.kt:62]: This better be a companion method?

File: AnkiDroid/src/main/java/com/ichi2/anki/services/ServiceConnection.kt [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+            context.bindService(intent, connection, Context.BIND_AUTO_CREATE)
+            serviceBound = true
+        }
+    }
+
+    open fun onServiceConnected(service: T) {}
+    open fun onServiceDisconnected(service: T) {}
+}
+
+interface SimpleBinder<T : Service> {
```
[the remark this answers is not in the document]
Comment (oakkitten) [reply AnkiDroid/src/main/java/com/ichi2/anki/services/ServiceConnection.kt:92]: This is marked as resolved but I don't see how this javadoc is helpful at all.

File: AnkiDroid/src/main/java/com/ichi2/utils/FileUtil.kt [hunk 10]
```diff
@@ -81,6 +81,27 @@ object FileUtil {
...
+     *
+     * @param file Abstract representation of the file/directory whose size needs to be calculated
+     * @return Size of the File/Directory in bytes. 0 if the [File] does not exist
+     */
+    fun getSize(file: File): Long {
+        if (file.isFile) {
+            return file.length()
+        } else if (!file.exists()) {
+            return 0L
+        }
```
Comment (oakkitten) [inline AnkiDroid/src/main/java/com/ichi2/utils/FileUtil.kt:101]: This seems to be an overcomplication of `CoroutineScope.calculateSize(file: File)`. Use that?

File: AnkiDroid/src/main/java/com/ichi2/utils/HandlerUtils.kt [hunk 11]
```diff
@@ -84,3 +84,11 @@ object HandlerUtils {
         )
     }
 }
+
+fun runOnUiThread(runnable: () -> Unit) {
```
Comment (oakkitten) [inline AnkiDroid/src/main/java/com/ichi2/utils/HandlerUtils.kt:88]: I suggest consolidating this with stuff like `executeFunctionUsingHandler`, `executeFunctionWithDelay`, etc

File: AnkiDroid/src/main/java/com/ichi2/utils/Repeater.kt [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+import com.ichi2.utils.HandlerUtils.newHandler
+import timber.log.Timber
+
+/**
+ * Schedules repeating events
+ * @param delayMs the interval to wait after the execution of [runnable] before the next call is made
+ * @param runnable Called once every [delayMs] until [terminate] is called
+ */
+class Repeater private constructor(val delayMs: Long, val runnable: () -> Unit) {
```
Comment (oakkitten) [inline AnkiDroid/src/main/java/com/ichi2/utils/Repeater.kt:27]: I'd suggest replacing this class with a coroutine, or removing the logic entirely in favor of `StateFlow` and friends

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt [hunk 13]
```diff
@@ -2779,31 +2852,46 @@ open class DeckPicker :
...
     }
 
     /**
      * Show a dialog that explains no sync can occur during migration.
      */
     private fun warnNoSyncDuringMigration() {
-        // TODO: Fetch and display real numbers
+        // TODO: handle value updates
+        // Note: migrationService shouldn't be null in normal operation
+        val text = migrationService.instance?.let { service ->
```
[the remark this answers is not in the document]
Comment (criticalAY) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:2886]: This function containing the code you mention is temporary as it has been replaced in my PR that handles the progress bar during migration so the lines inside this particular function need not to be worried about
